tcp/client.go

```diff
@@ -433,59 +423,60 @@ func (client *Client) dial(ctx context.Context, addr string) (net.Conn, handshak
 	dialCtx, dialCancel := context.WithTimeout(ctx, client.opts.TimeToLive)
 	defer dialCancel()
 
-	// Remember the number of dial attempts for this connection. If this
-	// exceeds MaxDialAttempts, the connection will be dropped.
+	// Remember the number of dial attempts for this connection. If this exceeds
+	// MaxDialAttempts, the connection will be dropped.
 	numDialAttempts := 0
 	for {
 		select {
 		case <-dialCtx.Done():
-			// dial must only return an error if it cannot successful
-			// dial within the TimeToLive duration. Othewise, it will keep
-			// retrying.
+			// dial must only return an error if it cannot successful dial
+			// within the TimeToLive duration. Othewise, it will keep retrying.
 			return nil, nil, dialCtx.Err()
 		default:
 		}
 
 		// Increment the number of dial attempts.
 		numDialAttempts++
 
+		// If the number of dial attempts has exceeded the maximum, return an
+		// error.
+		if numDialAttempts > client.opts.MaxDialAttempts {
+			return nil, nil, fmt.Errorf("dialing: exceeded max dial attempts")
+		}
+
 		// Attempt to dial a new connection for TimeToDial seconds. If we are
 		// not successful, then wait until the end of the TimeToDial second
 		// timeout and try again.
 		innerDialCtx, innerDialCancel := context.WithTimeout(dialCtx, client.opts.TimeToDial)
```

Comment:
Consider backing-off exponentially (max-ing out at some configurable `TimeToDialMax` option).

Reply:
@loongy I've added a configurable `TimeToDialBackoff` option instead since it provides a bit more customisability (without introducing much more logic).